src/clients/java/src/zig/src/client.zig

```diff
@@ -156,128 +187,205 @@ const RequestReflection = struct {
+        tb.tb_client_deinit(context.client);
     }
 
-    /// JNI Client.submit native implementation
+    /// JNI submit implementation.
     fn submit(
         env: *jui.JNIEnv,
+        context: *JNIContext,
         request_obj: jui.jobject,
-        client: tb.tb_client_t,
-        packet: *tb.tb_packet_t,
     ) void {
         assert(request_obj != null);
 
-        // Holds a global reference to prevent GC during the callback
+        // Holds a global reference to prevent GC during the callback.
         var global_ref = env.newReference(.global, request_obj) catch {
-            // NewGlobalRef fails only when the JVM runs out of memory
-            @panic("JNI: Error creating a global reference");
+            // NewGlobalRef fails only when the JVM runs out of memory.
+            @panic("JNI: Unexpected error creating a global reference");
         };
 
         assert(global_ref != null);
         errdefer env.deleteReference(.global, global_ref);
 
-        var buffer = RequestReflection.buffer(env, request_obj) orelse {
+        var buffer = ReflectionHelper.buffer(env, request_obj) orelse {
             // It is unexpected to the buffer be null here
             // The java side must allocate a new buffer prior to invoking "submit".
             @panic("JNI: Request buffer cannot be null");
         };
 
-        packet.operation = RequestReflection.operation(env, request_obj);
+        var packet = context.acquire_packet() orelse {
+            // It is unexpected to not have any packet available here.
+            // The java side syncronize how many threads can access.
+            @panic("JNI: No available packets");
+        };
+        packet.operation = ReflectionHelper.operation(env, request_obj);
         packet.user_data = global_ref;
         packet.data = buffer.ptr;
         packet.data_size = @intCast(u32, buffer.len);
         packet.next = null;
         packet.status = .ok;
 
         var packet_list = tb.tb_packet_list_t.from(packet);
-        tb.tb_client_submit(client, &packet_list);
+        tb.tb_client_submit(context.client, &packet_list);
     }
 
-    /// Completion callback
+    /// Completion callback.
     fn on_completion(
-        context: usize,
+        context_ptr: usize,
         client: tb.tb_client_t,
         packet: *tb.tb_packet_t,
         result_ptr: ?[*]const u8,
         result_len: u32,
     ) callconv(.C) void {
         _ = client;
 
-        var jvm = @intToPtr(*jui.JavaVM, context);
-        var env = jvm.attachCurrentThreadAsDaemon() catch |err| {
-            log.err("Error {s} attaching the native thread as daemon", .{@errorName(err)});
-            @panic("JNI: Error attaching the native thread as daemon");
+        var context = @intToPtr(*JNIContext, context_ptr);
+        var env = context.jvm.attachCurrentThreadAsDaemon() catch |err| {
+            log.err("Unexpected error attaching the native thread as daemon {s}", .{err});
```

Comment:
I'm not sure why the zig compiler isn't complaining, but shouldn't this just be `{}` for formatting now that the error is being printed directly (i.e. without `@errorName()`)?

Reply:
That's right, I replaced some occurrences of `"{s}", .{ @errorName(err) }` with `"{}", .{ err }` and I must have mixed this up.

Idk why it compiled either.